src/helpers/offerings-parser.ts

```diff
@@ -0,0 +1,158 @@
+  if (offeringIdsByPlacement == null) {
+    return null;
+  }
+  const fallbackOfferingId = placementsData.fallback_offering_id ?? null;
+  let offering: Offering | undefined;
+  if (placementId in offeringIdsByPlacement) {
+    const offeringId = offeringIdsByPlacement[placementId] ?? null;
+    if (offeringId == null) {
+      return null;
+    }
+    offering = allOfferings[offeringId];
+    if (offering == undefined) {
+      offering = allOfferings[fallbackOfferingId];
+    }
+  } else {
+    offering = allOfferings[fallbackOfferingId];
+  }
+
+  if (offering == undefined) {
+    return null;
+  }
+
+  return {
+    ...offering,
+    packagesById: Object.fromEntries(
+      Object.entries(offering.packagesById).map(([packageId, rcPackage]) => [
+        packageId,
+        addPlacementContextToPackage(rcPackage, placementId),
+      ]),
+    ),
+    availablePackages: offering.availablePackages.map((rcPackage) =>
+      addPlacementContextToPackage(rcPackage, placementId),
+    ),
+    weekly: addPlacementContextToNullablePackage(offering.weekly, placementId),
+    monthly: addPlacementContextToNullablePackage(
+      offering.monthly,
+      placementId,
+    ),
+    twoMonth: addPlacementContextToNullablePackage(
+      offering.twoMonth,
+      placementId,
+    ),
+    threeMonth: addPlacementContextToNullablePackage(
+      offering.threeMonth,
+      placementId,
+    ),
+    sixMonth: addPlacementContextToNullablePackage(
+      offering.sixMonth,
+      placementId,
+    ),
+    annual: addPlacementContextToNullablePackage(offering.annual, placementId),
+    lifetime: addPlacementContextToNullablePackage(
+      offering.lifetime,
+      placementId,
+    ),
+  };
+};
+
+export function toOfferings(
+  offeringsData: OfferingsResponse,
+  productsData: ProductsResponse,
+): Offerings {
+  const productsMap: { [productId: string]: ProductResponse } = {};
+  productsData.product_details.forEach((p: ProductResponse) => {
+    productsMap[p.identifier] = p;
+  });
+
+  const allOfferings: { [offeringId: string]: Offering } = {};
+  offeringsData.offerings.forEach((o: OfferingResponse) => {
+    const isCurrent = o.identifier === offeringsData.current_offering_id;
+    const offering = toOffering(
+      isCurrent,
+      o,
+      productsMap,
+      offeringsData.targeting,
+    );
+    if (offering != null) {
+      allOfferings[o.identifier] = offering;
+    }
+  });
+
+  const currentOffering: Offering | null = offeringsData.current_offering_id
+    ? allOfferings[offeringsData.current_offering_id] ?? null
+    : null;
+
+  if (Object.keys(allOfferings).length == 0) {
+    Logger.debugLog(
+      "Empty offerings. Please make sure you've configured offerings correctly in the " +
+        "RevenueCat dashboard and that the products are properly configured.",
+    );
+  }
+
+  return {
+    all: allOfferings,
+    current: currentOffering,
+    getCurrentOfferingByPlacementId(placementId: string): Offering | null {
```

Comment:
Why don't we pre-initialise a `byPlacementId: {[id:string]:Offering}` and pre-calculate it here?

Then anyone can just access it with:
```
offerings.byPlacementId[placementId]
```

Reply:
Well, it might make sense to do that, we will just increase the size of the offerings object I guess, but it would certainly be cleaner... cc @joshdholtz since you worked on the mobile SDK for these I think, in case you have any thoughts on this approach!

Reply:
Answered in Slack but there is fallback logic where using any placement identifier should return  fallback offering. That doesn't work well with a dictionary because a placement that doesn't exist would return null 🤷‍♂️